File: common/src/main/java/com/dreamdisplays/Initializer.java
package com.dreamdisplays;

import com.dreamdisplays.downloader.Init;
import com.dreamdisplays.net.Packets.*;
import com.dreamdisplays.screen.Configuration;
import com.dreamdisplays.screen.Manager;
import com.dreamdisplays.screen.Screen;
import com.dreamdisplays.screen.Settings;
import com.dreamdisplays.util.Facing;
import com.dreamdisplays.util.RayCasting;
import com.dreamdisplays.util.Utils;
import me.inotsleep.utils.logging.LoggingManager;
import net.minecraft.client.Minecraft;
import net.minecraft.client.multiplayer.ClientLevel;
import net.minecraft.network.protocol.common.custom.CustomPacketPayload;
import net.minecraft.world.effect.MobEffectInstance;
import net.minecraft.world.effect.MobEffects;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.phys.BlockHitResult;
import org.joml.Vector3i;
import org.jspecify.annotations.NullMarked;
import org.jspecify.annotations.Nullable;
import org.lwjgl.glfw.GLFW;
import org.slf4j.LoggerFactory;

import java.io.File;
import java.util.Objects;
import java.util.UUID;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.concurrent.atomic.AtomicReference;

/**
 * Main initializer.
 */
@NullMarked
public class Initializer {

    public static final String MOD_ID = "dreamdisplays";
    private static final boolean[] wasPressed = {false};
    private static final AtomicBoolean wasInMultiplayer = new AtomicBoolean(
            false
    );
    private static final AtomicReference<@Nullable ClientLevel> lastLevel =
            new AtomicReference<>(null);
    private static final AtomicBoolean wasFocused = new AtomicBoolean(false);
    public static Config config = new Config(new File("./config/" + MOD_ID));
    public static Thread timerThread = new Thread(() -> {
        int lastDistance = 64;
        boolean isErrored = false;
        while (!isErrored) {
            Manager.getScreens().forEach(Screen::reloadQuality);
            if (config.defaultDistance != lastDistance) {
                config.defaultDistance = lastDistance;
                config.save();
            }
            try {
                Thread.sleep(2500);
            } catch (InterruptedException e) {
                isErrored = true;
            }
        }
    });
    public static boolean isOnScreen = false;
    public static boolean focusMode = false;
    public static boolean displaysEnabled = true;
    public static boolean isPremium = false;
    public static boolean isReportingEnabled = true;
    private static @Nullable Screen hoveredScreen = null;
    private static Mod mod;

    public static Config getConfig() {
        return config;
    }

    public static void onModInit(Mod dreamDisplaysMod) {
        mod = dreamDisplaysMod;
        LoggingManager.setLogger(LoggerFactory.getLogger(MOD_ID));
        LoggingManager.info("Starting Dream Displays...");
        config.reload();

        // Load client display settings
        Settings.load();

        Init.init();
        new Focuser().start();

        timerThread.start();
    }

    public static void onDisplayInfoPacket(Info packet) {
        if (!Initializer.displaysEnabled) return;

        if (Manager.screens.containsKey(packet.uuid())) {
            Screen screen = Manager.screens.get(packet.uuid());
            screen.updateData(packet);
            return;
        }

        createScreen(
                packet.uuid(),
                packet.ownerUuid(),
                packet.pos(),
                packet.facing(),
                packet.width(),
                packet.height(),
                packet.url(),
                packet.lang(),
                packet.isSync()
        );
    }

    public static void onDisplayEnabledPacket(DisplayEnabled packet) {
        Initializer.displaysEnabled = packet.enabled();
        config.displaysEnabled = packet.enabled();
        config.save();
    }

    public static void createScreen(
            UUID uuid,
            UUID ownerUuid,
            Vector3i pos,
            Facing facing,
            int width,
            int height,
            String code,
            String lang,
            boolean isSync
    ) {
        Screen screen = new Screen(
                uuid,
                ownerUuid,
                pos.x(),
                pos.y(),
                pos.z(),
                facing.toString(),
                width,
                height,
                isSync
        );

//        Settings.FullDisplayData savedData = Settings.getDisplayData(uuid);
//        int renderDistance = savedData != null ? savedData.renderDistance : config.defaultDistance;
//        screen.setRenderDistance(renderDistance);

        Player player = Minecraft.getInstance().player;
        if (
                player != null &&
                        screen.getDistanceToScreen(player.blockPosition()) >
                                Initializer.config.defaultDistance
        ) return;
        Manager.registerScreen(screen);
        if (!Objects.equals(code, "")) screen.loadVideo(code, lang);
    }

    public static void onSyncPacket(Sync packet) {
        if (!Manager.screens.containsKey(packet.uuid())) return;
        Screen screen = Manager.screens.get(packet.uuid());
        if (screen != null) {
            screen.updateData(packet);
        }
    }

    private static void checkVersionAndSendPacket() {
        try {
            String version = Utils.getModVersion();
            sendPacket(new Version(version));
        } catch (Exception e) {
            LoggingManager.error("Unable to get version", e);
        }
    }

    public static void onEndTick(Minecraft minecraft) {
        ClientLevel level = minecraft.level;
        if (level != null && minecraft.getCurrentServer() != null) {
            if (lastLevel.get() == null) {
                lastLevel.set(level);
                checkVersionAndSendPacket();
            }

            if (level != lastLevel.get()) {
                lastLevel.set(level);

                Manager.unloadAll();
                hoveredScreen = null;

                checkVersionAndSendPacket();
            }

            wasInMultiplayer.set(true);
        } else {
            if (wasInMultiplayer.get()) {
                wasInMultiplayer.set(false);
                Manager.unloadAll();
                hoveredScreen = null;
                lastLevel.set(null);
                return;
            }
        }

        BlockHitResult result = RayCasting.rCBlock(64);
        hoveredScreen = null;
        Initializer.isOnScreen = false;
        Player player = minecraft.player;
        if (player == null) return;
        for (Screen screen : Manager.getScreens()) {
            double displayRenderDistance = screen.getRenderDistance();

            if (
                    displayRenderDistance <
                            screen.getDistanceToScreen(player.blockPosition()) ||
                            !Initializer.displaysEnabled
            ) {
                Manager.saveScreenData(screen);
                Manager.unregisterScreen(screen);
                if (hoveredScreen == screen) {
                    hoveredScreen = null;
                    Initializer.isOnScreen = false;
                }
            } else {
                if (result != null) if (
                        screen.isInScreen(result.getBlockPos())
                ) {
                    hoveredScreen = screen;
                    Initializer.isOnScreen = true;
                }

                screen.tick(player.blockPosition());
            }
        }

        long window = minecraft.getWindow().handle();
        boolean pressed =
                GLFW.glfwGetMouseButton(window, GLFW.GLFW_MOUSE_BUTTON_RIGHT) ==
                        GLFW.GLFW_PRESS;

        if (pressed && !wasPressed[0]) {
            if (player.isShiftKeyDown()) {
                checkAndOpenScreen();
            }
        }

        wasPressed[0] = pressed;

        if (Initializer.focusMode && hoveredScreen != null) {
            player.addEffect(
                    new MobEffectInstance(
                            MobEffects.BLINDNESS,
                            20 * 2,
                            1,
                            false,
                            false,
                            false
                    )
            );

            wasFocused.set(true);
        } else if (!Initializer.focusMode && wasFocused.get()) {
            player.removeEffect(MobEffects.BLINDNESS);
            wasFocused.set(false);
        }
    }

    private static void checkAndOpenScreen() {
        if (hoveredScreen == null) return;
        Configuration.open(hoveredScreen);
    }

    public static void sendPacket(CustomPacketPayload packet) {
        mod.sendPacket(packet);
    }

    public static void onDeletePacket(Delete packet) {
        Screen screen = Manager.screens.get(packet.uuid());
        if (screen != null) {
            Manager.unregisterScreen(screen);
        }

        Settings.removeDisplay(packet.uuid());
        LoggingManager.info(
                "Display deleted and removed from saved data: " + packet.uuid()
        );
    }

    public static void onStop() {
        Manager.saveAllScreens();
        timerThread.interrupt();
        Manager.unloadAll();
        Focuser.instance.interrupt();
    }

    public static void onPremiumPacket(Premium packet) {
        isPremium = packet.premium();
    }

    public static void onReportEnabledPacket(ReportEnabled packet) {
        isReportingEnabled = packet.enabled();
    }
}

File: common/src/main/java/com/dreamdisplays/Initializer.kt
package com.dreamdisplays

import com.dreamdisplays.net.*
import com.dreamdisplays.screen.Manager
import com.dreamdisplays.screen.Menu
import com.dreamdisplays.screen.Screen
import com.dreamdisplays.screen.Settings
import com.dreamdisplays.util.Facing
import com.dreamdisplays.util.RayCasting
import com.dreamdisplays.util.Utils
import me.inotsleep.utils.logging.LoggingManager
import net.minecraft.client.Minecraft
import net.minecraft.client.multiplayer.ClientLevel
import net.minecraft.network.protocol.common.custom.CustomPacketPayload
import net.minecraft.world.effect.MobEffectInstance
import net.minecraft.world.effect.MobEffects
import org.bytedeco.ffmpeg.global.*
import org.bytedeco.javacpp.Loader
import org.joml.Vector3i
import org.jspecify.annotations.NullMarked
import org.lwjgl.glfw.GLFW
import org.slf4j.LoggerFactory
import java.io.File
import java.util.*
import java.util.concurrent.atomic.AtomicBoolean
import java.util.concurrent.atomic.AtomicReference

@NullMarked
object Initializer {
    const val MOD_ID: String = "dreamdisplays"
    private val wasPressed = booleanArrayOf(false)
    private val wasInMultiplayer = AtomicBoolean(false)
    private val lastLevel = AtomicReference<ClientLevel?>(null)
    private val wasFocused = AtomicBoolean(false)
    @JvmField
    var config: Config = Config(File("./config/$MOD_ID"))
    var timerThread: Thread = Thread {
        val lastDistance = 64
        var isErrored = false
        while (!isErrored) {
            Manager.getScreens().forEach { obj: Screen? -> obj!!.reloadQuality() }
            if (config.defaultDistance != lastDistance) {
                config.defaultDistance = lastDistance
                config.save()
            }
            try {
                Thread.sleep(2500)
            } catch (_: InterruptedException) {
                isErrored = true
            }
        }
    }
    @JvmField
    var isOnScreen: Boolean = false
    var focusMode: Boolean = false
    @JvmField
    var displaysEnabled: Boolean = true
    var isPremium: Boolean = false
    private var hoveredScreen: Screen? = null
    private var mod: Mod? = null

    @JvmStatic
    fun onModInit(dreamDisplaysMod: Mod) {
        mod = dreamDisplaysMod
        LoggingManager.setLogger(LoggerFactory.getLogger(MOD_ID))
        LoggingManager.info("Starting Dream Displays")
        config.reload()

        // Load client display settings
        Settings.load()

        // Initialize FFmpeg native libraries
        try {
            LoggingManager.info("Loading FFmpeg native libraries...")
            Loader.load(avutil::class.java)
            Loader.load(avcodec::class.java)
            Loader.load(avformat::class.java)
            Loader.load(swscale::class.java)
            Loader.load(swresample::class.java)
            LoggingManager.info("FFmpeg libraries loaded successfully!")
        } catch (e: Exception) {
            LoggingManager.error("Failed to load FFmpeg libraries", e)
        }

        Focuser().start()

        timerThread.start()
    }

    @JvmStatic
    fun onDisplayInfoPacket(packet: Info) {
        if (!displaysEnabled) return

        if (Manager.screens.containsKey(packet.id)) {
            val screen: Screen = Manager.screens.get(packet.id)!!
            screen.updateData(packet)
            return
        }

        createScreen(
            packet.id,
            packet.ownerId,
            packet.pos,
            packet.facing,
            packet.width,
            packet.height,
            packet.url,
            packet.lang,
            packet.isSync
        )
    }

    fun createScreen(
        id: UUID,
        ownerId: UUID,
        pos: Vector3i,
        facing: Facing,
        width: Int,
        height: Int,
        code: String,
        lang: String,
        isSync: Boolean
    ) {
        val screen = Screen(id, ownerId, pos.x(), pos.y(), pos.z(), facing.toString(), width, height, isSync)
        checkNotNull(Minecraft.getInstance().player)
        if (screen.getDistanceToScreen(Minecraft.getInstance().player!!.blockPosition()) > config.defaultDistance) return
        Manager.registerScreen(screen)
        if (code != "") screen.loadVideo(code, lang)
    }

    @JvmStatic
    fun onSyncPacket(packet: Sync) {
        if (!Manager.screens.containsKey(packet.id)) return
        val screen = Manager.screens.get(packet.id)
        screen?.updateData(packet)
    }

    private fun checkVersionAndSendPacket() {
        try {
            val version = Utils.modVersion
            sendPacket(Version(version))
        } catch (e: Exception) {
            LoggingManager.error("Unable to get version", e)
        }
    }

    @JvmStatic
    fun onEndTick(minecraft: Minecraft) {
        if (minecraft.level != null && minecraft.currentServer != null) {
            if (lastLevel.get() == null) {
                lastLevel.set(minecraft.level)
                checkVersionAndSendPacket()
            }

            if (minecraft.level !== lastLevel.get()) {
                lastLevel.set(minecraft.level)

                Manager.unloadAll()
                hoveredScreen = null

                checkVersionAndSendPacket()
            }

            wasInMultiplayer.set(true)
        } else {
            if (wasInMultiplayer.get()) {
                wasInMultiplayer.set(false)
                Manager.unloadAll()
                hoveredScreen = null
                lastLevel.set(null)
                return
            }
        }

        if (minecraft.player == null) return

        val result = RayCasting.rCBlock(64.0)
        hoveredScreen = null
        isOnScreen = false

        for (screen in Manager.getScreens()) {
            if (config.defaultDistance < screen.getDistanceToScreen(minecraft.player!!.blockPosition()) || !displaysEnabled) {
                Manager.unregisterScreen(screen)
                if (hoveredScreen === screen) {
                    hoveredScreen = null
                    isOnScreen = false
                }
            } else {
                if (result != null) if (screen.isInScreen(result.blockPos)) {
                    hoveredScreen = screen
                    isOnScreen = true
                }

                screen.tick(minecraft.player!!.blockPosition())
            }
        }

        val window = minecraft.window.handle()
        val pressed = GLFW.glfwGetMouseButton(window, GLFW.GLFW_MOUSE_BUTTON_RIGHT) == GLFW.GLFW_PRESS

        if (pressed && !wasPressed[0]) {
            if (minecraft.player != null && minecraft.player!!.isShiftKeyDown) {
                checkAndOpenScreen()
            }
        }

        wasPressed[0] = pressed

        if (focusMode && minecraft.player != null && hoveredScreen != null) {
            minecraft.player!!.addEffect(
                MobEffectInstance(
                    MobEffects.BLINDNESS,
                    20 * 2,
                    1,
                    false,
                    false,
                    false
                )
            )

            wasFocused.set(true)
        } else if (!focusMode && wasFocused.get() && minecraft.player != null) {
            minecraft.player!!.removeEffect(MobEffects.BLINDNESS)
            wasFocused.set(false)
        }
    }

    private fun checkAndOpenScreen() {
        if (hoveredScreen == null) return
        Menu.open(hoveredScreen!!)
    }

    @JvmStatic
    fun sendPacket(packet: CustomPacketPayload) {
        mod!!.sendPacket(packet)
    }

    @JvmStatic
    fun onDeletePacket(deletePacket: Delete) {
        val screen = Manager.screens.get(deletePacket.id) ?: return

        Manager.unregisterScreen(screen)
    }

    @JvmStatic
    fun onStop() {
        timerThread.interrupt()
        Manager.unloadAll()
        Focuser.instance.interrupt()
    }

    @JvmStatic
    fun onPremiumPacket(packet: Premium) {
        isPremium = packet.premium
    }
}
